### ISSUE TITLE
Reversing Mpow flame S

### ISSUE THREAD

**Issue by Ph0rk0z** [TheWeirdDev/Bluetooth_Headset_Battery_Level#21]
I have an MPow flame 2 and a flame S now. The S shows battery level in android. How can figure out the commands needed to make it work? Script always returns connection refused or device busy on all ports.

**Comment by TheWeirdDev** [comment]
There's a comment [here](https://github.com/TheWeirdDev/Bluetooth_Headset_Battery_Level/issues/16#issuecomment-626850818) about reversing the protocol using android devices.

You can figure out what `AT Command`s are sent and received but in order to make it work, your device needs to connect first. If it doesn't connect on any port, we can't use those commands.

Try waiting 5 seconds before trying another port. I think some ports block the device temporarily. That might be a reason you are getting `device busy` error.

**Comment by Ph0rk0z** [comment]
It has a serial port you can connect to but it never does that by itself.

**Comment by Ph0rk0z** [comment]
Ok, here is the magic happening:

```
   (Mpow Flame S)          HFP      26     Rcvd AT+BRSF=959 
   (Samsung Galaxy S III)  HFP      29     Sent   +BRSF: 3431  
   (Samsung Galaxy S III)  HFP      19     Sent   OK  
   (Mpow Flame S)          HFP      23     Rcvd AT+CIND=? 
   (Samsung Galaxy S III)  HFP      147    Sent   +CIND: ("CALL",(0,1)),("CALLSETUP",(0-3)),("SERVICE",(0-1)),("SIGNAL",(0-5)),("ROAM",(0,1)),("BATTCHG",(0-5)),("CALLHELD",(0-2))  
   (Samsung Galaxy S III)  HFP      19     Sent   OK  
   (Mpow Flame S)          HFP      22     Rcvd AT+CIND? 
   (Samsung Galaxy S III)  HFP      38     Sent   +CIND: 0,0,1,4,0,1,0  
   (Samsung Galaxy S III)  HFP      19     Sent   OK  
   (Mpow Flame S)          HFP      32     Rcvd AT+CMER=3, 0, 0, 1 
   (Samsung Galaxy S III)  HFP      20     Sent   OK  
   (Mpow Flame S)          HFP      23     Rcvd AT+CHLD=? 
   (Samsung Galaxy S III)  HFP      34     Sent   +CHLD: (0,1,2,3)  
   (Samsung Galaxy S III)  HFP      19     Sent   OK  
   (Mpow Flame S)          HFP      25     Rcvd AT+BIND=1,2 
   (Samsung Galaxy S III)  HFP      20     Sent   OK  
   (Mpow Flame S)          HFP      23     Rcvd AT+BIND=? 
   (Samsung Galaxy S III)  HFP      30     Sent   +BIND: (1,2)  
   (Samsung Galaxy S III)  HFP      19     Sent   OK  
   (Mpow Flame S)          HFP      22     Rcvd AT+BIND? 
   (Samsung Galaxy S III)  HFP      28     Sent   +BIND: 1,1  
   (Samsung Galaxy S III)  HFP      27     Sent   +BIND: 2,1  
   (Samsung Galaxy S III)  HFP      19     Sent   OK  
   (Mpow Flame S)          HFP      23     Rcvd AT+NREC=0 
   (Samsung Galaxy S III)  HFP      20     Sent   OK  
   (Mpow Flame S)          HFP      23     Rcvd AT+VGS=09 
   (Samsung Galaxy S III)  HFP      20     Sent   OK  
   (Mpow Flame S)          HFP      23     Rcvd AT+CLIP=1 
   (Samsung Galaxy S III)  HFP      20     Sent   OK  
   (Mpow Flame S)          HFP      23     Rcvd AT+CCWA=1 
   (Samsung Galaxy S III)  HFP      20     Sent   OK  
   (Mpow Flame S)          HFP      22     Rcvd AT+BTRH? 
   (Samsung Galaxy S III)  HFP      23     Sent   ERROR  
   (Mpow Flame S)          HFP      38     Rcvd AT+XAPL=0000-0000-0100,7 
   (Samsung Galaxy S III)  HFP      20     Sent   OK  
   (Mpow Flame S)          HFP      38     Rcvd AT+IPHONEACCEV=2,1,5,2,0 
   (Samsung Galaxy S III)  HFP      20     Sent   OK  
   (Mpow Flame S)          HFP      21     Rcvd AT+CLCC 
   (Samsung Galaxy S III)  HFP      20     Sent   OK  
   (Mpow Flame S)          HFP      36     Rcvd AT+BIA=1,1,1,0,0,0,1,0 
   (Samsung Galaxy S III)  HFP      20     Sent   OK  
   (Mpow Flame S)          HFP      36     Rcvd AT+CSRSF=0,0,0,1,0,0,0 
   (Samsung Galaxy S III)  HFP      23     Sent   ERROR  
```

**Comment by superbonaci** [comment]
@Ph0rk0z I don't recommend trying directly with the python script, unless you know what you are doing. How I did it is to use an Android device to do the job for me, phone or tablet can help.

Looks like you have a Samsung Galaxy S3. Do you have it stock or did you install other roms like CyanogenMod? Do you have itb rooted? All of this is important because I need you to confirm, that under Developer options on the Android, you can enable `Bluetooth HCI snoop log`.

**Comment by Ph0rk0z** [comment]
Its rooted with a custom rom. The stuff I posted was from HCI snoop log.

**Comment by superbonaci** [comment]
Ok, but are you able to see the battery level on bluetooth settings once Mpow flame S are connected (percentage in number)? What custom rom are you using, LineageOS or other? Which android version it is?

**Comment by Ph0rk0z** [comment]
Yes it shows a battery level. Its 7.1... but I don't see how the specifics of the phone matter.

**Comment by superbonaci** [comment]
Because the battery level is saved to log file, which is a pcap. But maybe it's not stored to HPF protocol... You would have to attach the log file to take a look at it. If you send it to me with a private message I can try to take a look if I find string.

**Comment by Ph0rk0z** [comment]
That is the log file.... commands between headset and phone. Some of them show up in the script but for some reason the script doesn't work.

**Comment by superbonaci** [comment]
You need to properly take the HCI snoop log file. First delete it on the phone, while or airplane mode. Then enable bluetooth and connect the earphones until you can see battery level on the phone. Then disable bluetooth, then fetch the log file. Please send me the log file using a private message, I'll take a look at it. The stuff you posted is not related.

**Comment by Ph0rk0z** [comment]
not related eh? yet same AT commands appear in the script? That's the entire dump of HFP packets after battery level was displayed. All I did was take the macs out.

**Comment by superbonaci** [comment]
Not all headsets use AT commands, but if you don't want to cooperate don't waste our time.

**Comment by TheWeirdDev** [comment]
@superbonaci Please stop spamming.

**Comment by Ph0rk0z** [comment]
This headset looks like it uses AT commands. I checked BT GATT attributes and there are none. Afaik there is only serial and GATT when pulling BT battery. This script uses all AT. Also there is some patched branch of pulse that handles battery level which would be ideal but I've not messed with it yet. I already used a different set of modules to enable APTX/aptxHD.

Instead of repeating to me how to take a log when I rooted/custom flashed my phone maybe let me know what you're thinking and how that would even work with this script or another linux project that can return battery level. Spread knowledge not anger.
